src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MatchXpathCheck.java

```diff
@@ -0,0 +1,215 @@
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
+import com.puppycrawl.tools.checkstyle.xpath.AbstractNode;
+import com.puppycrawl.tools.checkstyle.xpath.RootNode;
+import net.sf.saxon.Configuration;
+import net.sf.saxon.om.Item;
+import net.sf.saxon.sxpath.XPathDynamicContext;
+import net.sf.saxon.sxpath.XPathEvaluator;
+import net.sf.saxon.sxpath.XPathExpression;
+import net.sf.saxon.trans.XPathException;
+
+/**
+ * <p>
+ * Evaluates Xpath query and report violation on all matching AST nodes. Current check allows
+ * user to implement custom checks using Xpath.
+ * Please read more about xpath syntax at <a href="https://www.w3schools.com/xml/xpath_syntax.asp">
+ * W3Schools Xpath Syntax</a>. Information regarding xpath functions can be found at
+ * <a href="https://developer.mozilla.org/en-US/docs/Web/XPath/Functions">XSLT/XPath Reference</a>.
+ * If xpath query is not specified explicitly, then check does nothing.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code query} - Xpath query.
+ * Type is {@code java.lang.String}.
+ * Default value is {@code ""}.
+ * </li>
+ * </ul>
+ * <p>
+ * The following example demonstrates validation of methods order, so that public methods should
+ * come before the private ones:
+ * </p>
+ * <pre>
+ * &lt;module name="MatchXpath"&gt; &lt;property name="query"
+ * value="//METHOD_DEF[.//LITERAL_PRIVATE and
+ * following-sibling::METHOD_DEF[.//LITERAL_PUBLIC]]"/&gt; &lt;/module&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * public class Test {
+ *  public void method1() { }
+ *  private void method2() { } // violation
+ *  public void method3() { }
+ *  private void method4() { } // violation
+ *  public void method5() { }
+ *  private void method6() { } // ok
+ * }
+ * </pre>
+ * <p>
+ * To fail with violation if there are any parametrized constructors
+ * </p>
+ * <pre>
+ * &lt;module name="MatchXpath"&gt; &lt;property name="query"
+ * value="//CTOR_DEF[count(./PARAMETERS/node()) &gt; 0]"/&gt; &lt;/module&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * public class Test {
+ *  public Test(Object c) { } // violation
+ *  public Test(int a, HashMap&lt;String, Integer&gt; b) { } // violation
+ *  public Test() { } // ok
+ * }
+ * </pre>
+ * <p>
+ * To fail with violation if method name is 'test' or 'foo'
```

Comment:
`To violate ...`

we do not fail anything, we just put some severity of events and plugin configuration usually do decision to fail or .... .

Reply:
done